File: solution/054SpiralMatrix/054SpiralMatrix.go
package main

import "fmt"

func spiralOrder(matrix [][]int) []int {
	if matrix == nil {
		return nil
	}
	M := len(matrix)
	if M == 0 {
		return nil
	}
	N := len(matrix[0])
	if N == 0 {
		return nil
	}
	result := make([]int, M*N)
	k := 0
	m := M
	n := N
	u := 0
	v := 0
	i := u
	j := v
	for k < M*N {
		for i < n {
			result[k] = matrix[j][i]
			i++
			k++
		}
		if k >= M*N {
			break
		}
		i--
		j++

		for j < m {
			result[k] = matrix[j][i]
			j++
			k++
		}
		if k >= M*N {
			break
		}
		i--
		j--

		for i >= u {
			result[k] = matrix[j][i]
			i--
			k++
		}
		if k >= M*N {
			break
		}
		i++
		j--

		for j >= v+1 {
			result[k] = matrix[j][i]
			j--
			k++
		}
		if k >= M*N {
			break
		}
		i++
		j++

		m--
		n--
		u++
		v++
	}

	return result
}

func main() {
	fmt.Printf("%v\n", spiralOrder([][]int{}))
	fmt.Printf("%v\n", spiralOrder([][]int{
		{1, 2, 3},
	}))
	fmt.Printf("%v\n", spiralOrder([][]int{
		{1, 2},
		{4, 5},
	}))
	fmt.Printf("%v\n", spiralOrder([][]int{
		{1, 2, 3},
		{4, 5, 6},
	}))
	fmt.Printf("%v\n", spiralOrder([][]int{
		{1, 2, 3},
		{4, 5, 6},
		{7, 8, 9},
	}))
	fmt.Printf("%v\n", spiralOrder([][]int{
		{1, 2, 3, 4},
		{10, 11, 12, 5},
		{9, 8, 7, 6},
	}))
}

File: src/054SpiralMatrix/054SpiralMatrix.rs
impl Solution {
    pub fn spiral_order(matrix: Vec<Vec<i32>>) -> Vec<i32> {
        let mut n = matrix.len() as i32;
        if n == 0 {
            return vec![];
        }
        let mut m = matrix[0].len() as i32;
        if m == 0 {
            return vec![];
        }
        
        let mut result = vec![];
        
        let (mut u, mut v, mut k) = (0 as i32, 0 as i32, m*n);
        
        while k!= 0{
            let (mut i, mut j) = (u, v);
            while i < m && k != 0 { result.push(matrix[j as usize][i as usize]); i+=1; k-=1;}
            j+=1;i-=1; m-=1;
            while j < n && k != 0 { result.push(matrix[j as usize][i as usize]); j+=1; k-=1;}
            i-=1;j-=1; n-=1;
            while i >= u && k != 0 { result.push(matrix[j as usize][i as usize]); i-=1; k-=1;}
            j-=1;i+=1; v+=1;
            while j >= v && k != 0 { result.push(matrix[j as usize][i as usize]); j-=1; k-=1;}
            u+=1;
        }
        
        return result;
    }
}
